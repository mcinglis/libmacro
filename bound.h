
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_BOUND_H
#define LIBMACRO_BOUND_H


#include <stdint.h>
#include <limits.h>

#include <libtypes/types.h>


// @public
#define MIN_BOUND( X ) \
    _Generic( ( X ), \
        bool: false, \
        char: CHAR_MIN, \
        signed char: SCHAR_MIN, \
        unsigned char: 0, \
        signed short: SHRT_MIN, \
        unsigned short: 0, \
        signed int: INT_MIN, \
        unsigned int: 0, \
        signed long: LONG_MIN, \
        unsigned long: 0, \
        signed long long: LLONG_MIN, \
        unsigned long long: 0, \
        default: INTMAX_MIN \
    )


// @public
#define MAX_BOUND( X ) \
    _Generic( ( X ), \
        bool:                   true, \
        char:                   CHAR_MAX, \
        signed char:            SCHAR_MAX, \
        unsigned char:          UCHAR_MAX, \
        signed short:           SHRT_MAX, \
        unsigned short:         USHRT_MAX, \
        signed int:             INT_MAX, \
        unsigned int:           UINT_MAX, \
        signed long:            LONG_MAX, \
        unsigned long:          ULONG_MAX, \
        signed long long:       LLONG_MAX, \
        unsigned long long:     ULLONG_MAX, \
        default:                INTMAX_MAX \
    )


#endif // ifndef LIBMACRO_BOUND_H

