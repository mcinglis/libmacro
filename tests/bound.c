
#include <stdio.h>
#include <sys/types.h>

#include "../bound.h"


int
main( void )
{

    printf( "min of `time_t`: %jd\n",
            ( intmax_t ) MIN_BOUND( ( time_t ) 0 ) );
    printf( "min of `off_t`: %jd\n",
            ( intmax_t ) MIN_BOUND( ( off_t ) 0 ) );
    printf( "min of `pid_t`: %jd\n",
            ( intmax_t ) MIN_BOUND( ( pid_t ) 0 ) );
    printf( "min of `dev_t`: %jd\n",
            ( intmax_t ) MIN_BOUND( ( dev_t ) 0 ) );
    printf( "min of `blkcnt_t`: %jd\n",
            ( intmax_t ) MIN_BOUND( ( dev_t ) 0 ) );
    printf( "min of `uid_t`: %jd\n",
            ( intmax_t ) MIN_BOUND( ( uid_t ) 0 ) );
    printf( "min of `size_t`: %jd\n",
            ( intmax_t ) MIN_BOUND( ( size_t ) 0 ) );
    printf( "min of `ssize_t`: %jd\n",
            ( intmax_t ) MIN_BOUND( ( ssize_t ) 0 ) );
    printf( "min of `intmax_t`: %jd\n",
            ( intmax_t ) MIN_BOUND( ( intmax_t ) 0 ) );
    printf( "min of `uintmax_t`: %jd\n",
            ( intmax_t ) MIN_BOUND( ( uintmax_t ) 0 ) );

    printf( "\n" );
    printf( "max of `time_t`: %ju\n",
            ( uintmax_t ) MAX_BOUND( ( time_t ) 0 ) );
    printf( "max of `off_t`: %ju\n",
            ( uintmax_t ) MAX_BOUND( ( off_t ) 0 ) );
    printf( "max of `pid_t`: %ju\n",
            ( uintmax_t ) MAX_BOUND( ( pid_t ) 0 ) );
    printf( "max of `dev_t`: %ju\n",
            ( uintmax_t ) MAX_BOUND( ( dev_t ) 0 ) );
    printf( "max of `blkcnt_t`: %ju\n",
            ( uintmax_t ) MAX_BOUND( ( dev_t ) 0 ) );
    printf( "max of `uid_t`: %ju\n",
            ( uintmax_t ) MAX_BOUND( ( uid_t ) 0 ) );
    printf( "max of `size_t`: %ju\n",
            ( uintmax_t ) MAX_BOUND( ( size_t ) 0 ) );
    printf( "max of `ssize_t`: %ju\n",
            ( uintmax_t ) MAX_BOUND( ( ssize_t ) 0 ) );
    printf( "max of `intmax_t`: %ju\n",
            ( uintmax_t ) MAX_BOUND( ( intmax_t ) 0 ) );
    printf( "max of `uintmax_t`: %ju\n",
            ( uintmax_t ) MAX_BOUND( ( uintmax_t ) 0 ) );

}

