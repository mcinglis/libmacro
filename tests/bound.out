min of `time_t`: -9223372036854775808
min of `off_t`: -9223372036854775808
min of `pid_t`: -2147483648
min of `dev_t`: 0
min of `blkcnt_t`: 0
min of `uid_t`: 0
min of `size_t`: 0
min of `ssize_t`: -9223372036854775808
min of `intmax_t`: -9223372036854775808
min of `uintmax_t`: 0

max of `time_t`: 9223372036854775807
max of `off_t`: 9223372036854775807
max of `pid_t`: 2147483647
max of `dev_t`: 18446744073709551615
max of `blkcnt_t`: 18446744073709551615
max of `uid_t`: 4294967295
max of `size_t`: 18446744073709551615
max of `ssize_t`: 9223372036854775807
max of `intmax_t`: 9223372036854775807
max of `uintmax_t`: 18446744073709551615
